### PR TITLE
distsql, metrics: refine distsql metrics.

### PR DESCRIPTION
**Pull request by coocood**
Add scan keys metrics and removed query count metrics due to misleading result.

### CODE REVIEW

File: distsql/distsql.go [hunk 1]
```diff
@@ -81,7 +82,8 @@ type selectResult struct {
 	selectResp *tipb.SelectResponse
 	respChkIdx int
 
-	scanCount int64
+	scanKeys     int64
+	partialCount int64
```
Comment (shenli) [inline distsql/distsql.go]: Add comment for this.

File: distsql/distsql.go [hunk 2]
```diff
@@ -221,6 +230,10 @@ func (r *selectResult) readRowsData(chk *chunk.Chunk) (err error) {
 // Close closes selectResult.
 func (r *selectResult) Close() error {
 	// Close this channel tell fetch goroutine to exit.
+	if r.scanKeys > 0 {
```
Comment (shenli) [inline distsql/distsql.go]: if r.scanKeys >= 0

File: distsql/stream.go [hunk 3]
```diff
@@ -147,6 +152,10 @@ func (r *streamResult) NextRaw(goCtx goctx.Context) ([]byte, error) {
 }
 
 func (r *streamResult) Close() error {
+	if r.scanKeys > 0 {
```
Comment (shenli) [inline distsql/stream.go]: >= 0 ?

File: distsql/distsql.go [hunk 4]
```diff
@@ -221,6 +230,10 @@ func (r *selectResult) readRowsData(chk *chunk.Chunk) (err error) {
...
+		metrics.DistSQLScanKeysHistogram.Observe(float64(r.scanKeys))
+		metrics.DistSQLPartialCountHistogram.Observe(float64(r.partialCount))
```
Comment (lamxTyler) [inline distsql/distsql.go]: `partialCount` is irrelevant with `scanKeys`?

**Comment by coocood** [comment]
@shenli @lamxTyler PTAL

File: distsql/distsql.go [hunk 5]
```diff
@@ -327,16 +330,6 @@ func SelectDAG(goCtx goctx.Context, ctx context.Context, kvReq *kv.Request, fiel
 
 // Analyze do a analyze request.
 func Analyze(ctx goctx.Context, client kv.Client, kvReq *kv.Request) (SelectResult, error) {
-	var err error
-	defer func() {
-		// Add metrics.
-		if err != nil {
-			queryCounter.WithLabelValues(queryFailed).Inc()
```
Comment (lamxTyler) [inline distsql/distsql.go]: Why is this removed?

File: distsql/distsql.go [hunk 6]
```diff
@@ -136,16 +138,20 @@ func (r *selectResult) Next(goCtx goctx.Context) (PartialResult, error) {
...
+		r.scanKeys = -1
 	}
+	r.partialCount++
 	return pr, errors.Trace(err)
 }
 
 // NextRaw returns the next raw partial result.
 func (r *selectResult) NextRaw(goCtx goctx.Context) ([]byte, error) {
 	re := <-r.results
+	r.partialCount++
```
Comment (lamxTyler) [inline distsql/distsql.go]: Should we set the `scanKeys` to `-1`, because it will never update the `scanKeys`?

File: distsql/stream.go [hunk 7]
```diff
@@ -147,6 +152,10 @@ func (r *streamResult) NextRaw(goCtx goctx.Context) ([]byte, error) {
 }
```
Comment (lamxTyler) [inline distsql/stream.go]: Should we also update the `partialCount` and `scanKeys`?

File: distsql/distsql.go (hunk 5 above)
Comment (coocood) [reply distsql/distsql.go]: It's not accurate on failed or success, the actual error will be returned when we read the result.
And it doesn't provide much useful information.

File: distsql/stream.go (hunk 7 above)
Comment (coocood) [reply distsql/stream.go]: We can't get scanKeys for the raw response.

File: distsql/distsql.go (hunk 5 above)
Comment (lamxTyler) [reply distsql/distsql.go]: Got it, and please update the description of this PR.

File: distsql/stream.go (hunk 7 above)
Comment (lamxTyler) [reply distsql/stream.go]: I mean we can set the `scanKeys` to -1 like the `selectResult` do.

**Comment by coocood** [comment]
@lamxTyler PTAL

**Comment by shenli** [comment]
LGTM
